_posts/2015-12-30-Python-Join-Tutorial1.md

```diff
@@ -0,0 +1,302 @@
+---
+title: Real World Hadoop - Real World Hadoop Guide for Python 
+layout: post
+description: Data processing pipelines with Spark
+topic: engineering
+author: matthew_rathbone
+coauthor:
+  name: Elena Akhmatova
+  link: https://ru.linkedin.com/pub/elena-akhmatova/3/877/266
+categories:
+- hadoop
+tags:
+- python
+- hadoop
+- hadoop-streaming
+---
+
+> This article is part of [my guide to map reduce frameworks][4] in which I implement a solution to a real-world problem in each of the most popular Hadoop frameworks.  
+>  
+> One of the articles in the guide [Hadoop Python MapReduce Tutorial for Beginners][13] has already introduced the reader to the basics of hadoop-streaming with Python. This is the next logical step in a quest to learn how to use Python in map reduce framework defined by Hadoop.  
+> 
+
+## The Problem
+
+Let me quickly restate the problem from [my original article][6].
+
+I have two datasets:
+
+1. User information (id, email, language, location)
+2. Transaction information (transaction-id, product-id, user-id, purchase-amount, item-description)
+
+Given these datasets, I want to find the number of unique locations in which each product has been sold. To do that, I need to join the two datasets together.
+
+Previously I have implemented this solution [in java][7], [with hive][8] and [with pig][9]. The java solution was ~500 lines of code, hive and pig were like ~20 lines tops.
+
+The [Python article][13] i mentioned earlier does not solve this problem. It has a focus on understanding the basics of working with Hadoop-streaming and Python. 
+
+## The Python Solution
+
+Again, this article is a follow up for my [earlier article][13] on [Python][1] that shows the basics of using Python for map reduce in Hadoop. Please refer to it to gather basic understanding on how to write mappers and reducers in Python for Hadoop. Here we focus on implementing join and count distinct by key in Python and will assume some preliminary understanding of hadoop-streaming. 
+
+## Demonstration Data
+
+As previously (and here we mean "versus" the solutions [in java][7], [with hive][8] and [with pig][9]) we use two datasets called `users` and `transactions`. 
+
+{% highlight bash %}
+users
+1	matthew@test.com	EN	US
+2	matthew@test2.com	EN	GB
+3	matthew@test3.com	FR	FR
+{% endhighlight %}
+
+and
+
+{% highlight bash %}
+transactions
+1	1	1	300	a jumper
+2	1	2	300	a jumper
+3	1	2	300	a jumper
+4	2	3	100	a rubber chicken
+5	1	3	300	a jumper
+{% endhighlight %}
+
+**This time though they will constitute the same input dataset!** Soon it will become clear why.
+
+{% highlight bash %}
+hdfs dfs -mkdir input
+
+hdfs dfs -put ./users.txt input
+hdfs dfs -put ./transactions.txt input
+{% endhighlight %}
+
+And for command line testing
```

Comment:
Ok, so I looked at some documentation and we do not need to join them into a single file, but can specify multiple inputs on the command line when we start the job.

eg:

`./streaming.jar -input transactions.txt -input users.txt ...`